### ISSUE TITLE
Change feature type names in `RecordData`

### ISSUE THREAD

**Issue by huzecong** [asyml/texar-pytorch#217]
The feature types in `RecordData` (`FixedLenFeature`, `FixedLenSequenceFeature`, and `VarLenFeature`) are directly borrowed from TensorFlow and might cause confusion for PyTorch users. Based on my discussion with @AvinashBukkittu yesterday, we think it might be worthy to also introduce a set of aliases for these types, and use them in the examples.

Here's the meaning behind each feature type, and what happens when we collate such features:
- `FixedLenFeature` are tensor features that have a fixed shape. These features are stacked together (`torch.stack(features, dim=0)`) while collating.
- `FixedLenSequenceFeature` represent a list of tensor features such that each list element have the same fixed shape. These features are padded and batched (`tx.data.padded_batch(features)`) while collating.
- `VarLenFeature` can be tensors of any shape, or other objects that are not tensors. These features are simply put into a list (`list(features)`) while collating.

I think we can rename the feature types to either:
- describe the object types of the input features...
	- `tensor`, `tensor_seq`, and `any`/`other`.
- ... or, describe how they're collated.
	- `tensor`, `pad_tensor`, and `list`.

But yeah, any suggestions are welcome.

---

Also, for the `feature_original_types` format:
https://github.com/asyml/texar-pytorch/blob/43967ee238a5da2e996f4f71644a940a86cad009/texar/torch/data/data/record_data.py#L444-L470

Since `VarLenFeature` can be any object, it does not really make sense to ask the user to write a `dtype`. Maybe we can also accept `None` here.

**Comment by ZhitingHu** [comment]
> The feature types in `RecordData` (`FixedLenFeature`, `FixedLenSequenceFeature`, and `VarLenFeature`) 

These are called `feature_len_type` in the doc
https://github.com/asyml/texar-pytorch/blob/43967ee238a5da2e996f4f71644a940a86cad009/texar/torch/data/data/record_data.py#L455

Are we going to change the name too? The overall field name `feature_original_types` could also affect the naming here. If it's kept not changed, the naming based on `describe how they're collated` is probably not suitable

**Comment by ZhitingHu** [comment]
If these names are only for indicating the shapes, I think sth like `"FixedLen*"` or `"FixedShape*"` is probably more explicit than `"tensor"` (which does not necessary have fixed shapes)?

**Comment by huzecong** [comment]
I guess these original names are more intuitive to previous TF users, but they might be confusing for a PyTorch-only user in several aspects:

1. In TF, the dataset can only contain tensors, while this is not the case for PyTorch. `Fixed*Feature` would assume the data are all tensors, but it is not reflected in the naming.
2. There's no clear distinction between `FixedLenSequenceFeature` and `VarLenFeature`. In TF we'd use a `None`-dimension to indicate a non-fixed dimension size, but there's no such counterpart in PyTorch.

Indeed the name `tensor` by itself is not good enough. How about these:
- `stacked_tensor`, `padded_tensor`, `list`
The advantages of these names would be:
1. They're explicit about whether the type must be a tensor.
2. They indicate how the features are collated.
3. They're much shorter to write.

**Comment by ZhitingHu** [comment]
A quick question: if a feature is a single int (e.g., a label), would `stacked_tensor` be used? -- If so, would it be a bit confusing for PyTorch users cuz for them, if I'm correct, `int` is not a tensor?

I'm good with the new naming, the thing is that if we use it, the name `feature_original_types` should also be changed, cuz `stacked_tensor` indicates the type **after** processing (e.g., collating) instead of the original type.

**Comment by huzecong** [comment]
If the feature is a single int, `stacked_tensor` could be used. Internally `np.stack` is called, so as long as arguments are acceptable by `np.stack`, they can be used. But either way, `np.stack` will always return a NumPy `ndarray`, so the collated output is guaranteed to be a tensor.

I think `feature_original_types` is still acceptable, because these can still be interpreted as the (more abstract) types of the feature. What might needs change is the name `feature_len_type`: we could just change it to `feature_type`.

**Comment by ZhitingHu** [comment]
Yes, they are still "types" of the feature, but not "original". Perhaps we can rename it as `feature_types`, and `feature_len_type` as `feature_shape_type`?

For `stacked_tensor`, would `fixed_shape_tensor` clearer?

Either way, it's important to add in the doc what common types each "feature_type" can handle. For example, `stacked_tensor` is for single scalar, fixed shape features, etc

**Comment by huzecong** [comment]
I see. It makes more sense to call it `feature_types` instead of `features_original_types`, but this breaks compatibility with Texar-TF. What I wanted to achieve is that previous code that uses `features_original_types` and `**Feature` can still work (or maybe gradually deprecate them), but we encourage new code to use this different set of names.

If we decide on this set of names, which describes types of collated results, then I think it might be better to name this as `feature_collate_method` or something along the lines. Also, I think it's not really appropriate to use `fixed_shape_tensor` since it describes individual examples more, and is not as explicit compared to `stacked_tensor`.

**Comment by AvinashBukkittu** [comment]
My two cents on this is the following

- `feature_types` is more elegant than `feature_original_types`. We can mark `feature_original_types` as deprecated but still support it? Similarly, `feature_shape_type` is more explicit than `feature_len_type` (This naming change won't really affect our code base as we pass in a list, we just have to update the documentation)

- I like the following mapping
  `FixedLenFeature` -> `stacked_tensor`
  `FixedLenSequenceFeature` -> `stacked_padded_tensor` (If you feel this is verbose, `padded_tensor` is good)
  `VarLenFeature` -> `list`

**Comment by ZhitingHu** [comment]
`feature_collate_method` sounds good to me.

For other parts, I think Avinash made a good summarization.

**Comment by huzecong** [comment]
Cool, then I'll make a PR correspondingly.
